File: aether_flow_flutter/lib/clipboard_monitor.dart
import 'dart:async';
import 'package:flutter/services.dart';

/// 剪貼簿監控與操作 (Universal Clipboard 核心 - Flutter 實作)
class ClipboardMonitor {
  String _lastCopiedText = '';
  Timer? _timer;
  
  // Singleton pattern for easy access
  static final ClipboardMonitor _instance = ClipboardMonitor._internal();
  factory ClipboardMonitor() => _instance;
  ClipboardMonitor._internal();

  /// 開始監控剪貼簿
  void startMonitoring() {
    _timer = Timer.periodic(const Duration(milliseconds: 500), (timer) async {
      await _checkClipboard();
    });
  }

  /// 停止監控
  void stopMonitoring() {
    _timer?.cancel();
    _timer = null;
  }

  Future<void> _checkClipboard() async {
    try {
      final clipboardData = await Clipboard.getData(Clipboard.kTextPlain);
      final text = clipboardData?.text;
      
      if (text != null && text.isNotEmpty && text != _lastCopiedText) {
        _lastCopiedText = text;
        _onClipboardChanged(text);
      }
    } catch (e) {
      // 忽略部分平台讀取權限或錯誤
    }
  }

  /// 當本機剪貼簿改變時觸發
  void _onClipboardChanged(String newText) {
    print("📋 [Clipboard] Detected local copy: $newText");
    // TODO: 整合通訊層 (AetherCrypto 加密 + 網路發送)
  }

  /// 接收到來自其他設備的剪貼簿同步請求時呼叫
  Future<void> syncFromRemote(String text) async {
    _lastCopiedText = text;
    await Clipboard.setData(ClipboardData(text: text));
    print("🔄 [Clipboard] Synced from remote: $text");
    
    // TODO: 觸發本地系統通知 (可結合 flutter_local_notifications)
  }
}

File: aether_flow_apple/ClipboardMonitor.swift
import Foundation

#if os(macOS)
import AppKit
#else
import UIKit
#endif

/// 剪貼簿監控與操作 (Universal Clipboard 核心)
public class ClipboardMonitor: ObservableObject {
    @Published public var lastCopiedText: String = ""
    
    private var timer: Timer?
    private var lastChangeCount: Int = 0
    
    public init() {
        startMonitoring()
    }
    
    public func startMonitoring() {
        #if os(macOS)
        lastChangeCount = NSPasteboard.general.changeCount
        #endif
        
        timer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: true) { [weak self] _ in
            self?.checkClipboard()
        }
    }
    
    public func stopMonitoring() {
        timer?.invalidate()
        timer = nil
    }
    
    private func checkClipboard() {
        #if os(macOS)
        let currentCount = NSPasteboard.general.changeCount
        if currentCount != lastChangeCount {
            lastChangeCount = currentCount
            if let text = NSPasteboard.general.string(forType: .string) {
                DispatchQueue.main.async {
                    self.lastCopiedText = text
                    self.onClipboardChanged(text)
                }
            }
        }
        #else
        // iOS implementation using UIPasteboard
        if let text = UIPasteboard.general.string, text != lastCopiedText {
            DispatchQueue.main.async {
                self.lastCopiedText = text
                self.onClipboardChanged(text)
            }
        }
        #endif
    }
    
    /// 當本機剪貼簿改變時觸發，準備將內容透過 AetherCrypto 加密並發送至其他設備
    private func onClipboardChanged(_ newText: String) {
        print("📋 [Clipboard] Detected local copy: \(newText)")
        // TODO: 整合 mDNS 發現的設備，並透過 TCP/UDP 發送加密後的 Payload
    }
    
    /// 接收到來自其他設備的剪貼簿同步請求時呼叫
    public func syncFromRemote(text: String) {
        #if os(macOS)
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(text, forType: .string)
        lastChangeCount = NSPasteboard.general.changeCount
        #else
        UIPasteboard.general.string = text
        #endif
        
        DispatchQueue.main.async {
            self.lastCopiedText = text
            print("🔄 [Clipboard] Synced from remote: \(text)")
            // TODO: 發送本地系統通知 (NotificationCenter)
        }
    }
}
